File: java/com/countrygamer/countrygamercore/Base/client/gui/GuiButtonRedstone.java
package com.countrygamer.countrygamercore.base.client.gui;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.util.ResourceLocation;

import org.lwjgl.opengl.GL11;

import com.countrygamer.countrygamercore.common.Core;
import com.countrygamer.countrygamercore.common.lib.RedstoneState;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

/**
 * A new button which can detail one of three {@link RedstoneState}s.
 * 
 * @author Country_Gamer
 * 
 */
@SideOnly(Side.CLIENT)
public class GuiButtonRedstone extends GuiButton {
	
	private RedstoneState redstoneState;
	public boolean isActive = false;
	private boolean useOldStyle;
	
	public GuiButtonRedstone(int id, int x, int y, RedstoneState state) {
		this(id, x, y, state, false);
	}
	
	public GuiButtonRedstone(int id, int x, int y, RedstoneState state, boolean useOld) {
		super(id, x, y, 18, 18, "");
		this.redstoneState = state;
		this.useOldStyle = useOld;
	}
	
	@Override
	public void drawButton(Minecraft minecraft, int x, int y) {
		if (!this.enabled) this.isActive = false;
		if (this.visible) {
			minecraft.getTextureManager().bindTexture(
					new ResourceLocation(Core.pluginID, "textures/gui/buttons.png"));
			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
			
			int u = 100;
			int v = 0;
			
			int redstoneStateAsInt = RedstoneState.getIntFromState(this.redstoneState) + 1;
			u += redstoneStateAsInt * 18;
			
			if (this.useOldStyle) {
				u += 54;
			}
			
			boolean flag = x >= this.xPosition && y >= this.yPosition
					&& x < this.xPosition + this.width && y < this.yPosition + this.height;
			int hoverState = this.getHoverState(flag);
			v += hoverState * 18;
			
			if (this.isActive) v += 36;
			
			this.drawTexturedModalRect(this.xPosition, this.yPosition, u, v, this.width,
					this.height);
		}
	}
	
	
	public RedstoneState getRedstoneState() {
		return this.redstoneState;
	}
	
	public void nextRedstoneState() {
		this.redstoneState = RedstoneState.getStateFromInt(RedstoneState
				.getIntFromState(this.redstoneState) + 1);
	}
	
}

File: src/main/scala/com/countrygamer/cgo/client/gui/GuiButtonRedstone.scala
package com.countrygamer.cgo.client.gui

import com.countrygamer.cgo.common.Origin
import com.countrygamer.cgo.common.lib.RedstoneState
import com.countrygamer.cgo.common.lib.util.UtilRender
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiButton
import net.minecraft.util.ResourceLocation
import org.lwjgl.opengl.GL11

/**
 *
 *
 * @author CountryGamer
 */
@SideOnly(Side.CLIENT)
class GuiButtonRedstone(id: Int, x: Int, y: Int, val state: RedstoneState, val useOld: Boolean)
		extends GuiButton(id, x, y, 18, 18, "") {

	val texture: ResourceLocation = new
					ResourceLocation(Origin.pluginID, "textures/gui/buttons.png")
	// Default Constructor

	// End Constructor

	// Other Constructors
	def this(id: Int, x: Int, y: Int, state: RedstoneState) {
		this(id, x, y, state, false)
	}

	// End Constructors
	override def drawButton(minecraft: Minecraft, mouseX: Int, mouseY: Int): Unit = {
		if (this.visible) {
			UtilRender.bindResource(this.texture)
			GL11.glColor3f(1.0F, 1.0F, 1.0F)
			val isHoveredOn: Boolean = mouseX >= this.xPosition && mouseY >= this.yPosition &&
					mouseX < this.xPosition + this.width && mouseY < this.yPosition + this.height
			var u: Int = 100
			var v: Int = 0

			val redstoneStateAsInt: Int = RedstoneState.getIntFromState(this.state)
			u += redstoneStateAsInt * 18

			if (this.useOld) {
				u += 54
			}

			v += this.getHoverState(isHoveredOn) * 18

			if (this.enabled) v += 36

			this.drawTexturedModalRect(this.xPosition, this.yPosition, u, v, this.width,
				this.height)

		}
	}
}